### DISCUSSION TITLE
Foam wikilinks implementation

### DISCUSSION THREAD

**Discussion by apricot13** [foambubble/foam#1091] [Q&A]
I'm trying to understand the implementation of wikilinks in foam and if theres an option to help me out!

Foam assumes that all notes are never going to have duplicate filenames so if I have this folder structure:

```
- test.md
```

and I create a new note `[[test]]` works fine.

But if I then have this folder structure:

```
- test.md
- folder/test.md
```

`[[test]]` will give me an error

![image](https://user-images.githubusercontent.com/649148/197247385-439b0663-5d64-4b4f-8a32-3fd7e646ad53.png)

It feels like it would be safer to assume that there will always be a possibility of multiple notes with the same filename in subdirectories and default to the [unique identifier](https://github.com/foambubble/foam/blob/2385bd75b50b298e3cffc41cedc5545d3bb18114/docs/dev/proposals/wikilinks-in-foam.md) approach? 

This method seems like it would produce more work the larger your collection since you'll have to go back and fix a whole bunch of broken links constantly, and you have no control initially other than to manually add the `folder` part of `[[folder/test]]` the first time you create the wikilink.

Other than the `problems` tab I can't see a way to fix these issues in a nice way?
![image](https://user-images.githubusercontent.com/649148/197249317-bcde45fc-5815-43ce-8953-2d02079fb424.png)

I guess what I'm asking is is there an option to choose which wikilink style to support by default in foam? 

Just as an aside about wikilinks in general - is there a default or standard for wikilinks? Why has obsidian opted for the unique identifier as standard `[[folder/test]]`, foam for unique `[[test]]` and some other applications `[[Test note title]]`?

**Comment by riccardoferretti** [comment]
The idea of the unique identifier is to be "backward" compatible with wikilinks. 
Unless you run into the "issue" of having multiple notes with the same name, you don't even realise there is such a thing as a "unique identifier", as standard wikilinks just do the trick.

IIUC what you are suggesting, you would like wikilinks to immediately be in the identifier fashion.
This is something that could be done, but I guess what you mean is slightly different.
Currently unique identifiers are the default. The only thing is that `test` (in your example) _is_ the unique identifier until you add the second file.
One approach could be for Foam, whenever a new file is added or renamed, to automatically check if it changes the unique identifiers of other notes, in which case to automatically update those documents.

Another approach, which I think is what you were implying, is to always use a "full relative path" from the root, which is a valid unique identifier for a file. This could be achieved with the autocomplete, and possibly set via an option.
I thought about something along those lines in the past, but decided against it at the time, mostly because I wanted to see a problem before coming forward with a solution.

I hope this provides some context, let me know your thoughts
